crates/sp-lightclient/src/lib.rs

```diff
@@ -178,32 +173,29 @@ pub trait HeaderImporter<Header: HeaderT> {
             calculate_block_weight(&global_randomness.global_randomness, &pre_digest);
         let total_weight = parent_header.total_weight + block_weight;
 
-        // last best header should ideally be parent header. if not check for forks and prune accordingly
-        let last_best_header = Self::Storage::best_header();
-        if last_best_header.header.hash() != parent_header.header.hash() {
+        // last best header should ideally be parent header. if not check for forks and pick the best chain
+        let last_best_header = store.best_header();
+        let is_best_header = if last_best_header.header.hash() == parent_header.header.hash() {
+            // header is extending the current best header. consider this best header
+            true
+        } else {
             let last_best_weight = last_best_header.total_weight;
-            let should_prune_and_import = match total_weight.cmp(&last_best_weight) {
-                // current weight is greater than last best. pick this header
+            match total_weight.cmp(&last_best_weight) {
+                // current weight is greater than last best. pick this header as best
                 Ordering::Greater => true,
                 // if weights are equal, pick the longest chain
-                // current header will always be rejected as its improbable to be longer chain than
-                // the best imported header
                 Ordering::Equal => header.number() > last_best_header.header.number(),
                 // we already are on the best chain
                 Ordering::Less => false,
-            };
-
-            if !should_prune_and_import {
-                return Err(ImportError::NonCanonicalHeader);
             }
-
-            // re-org: prune all the descendants of the parent header and them import this header as best
-            Self::Storage::prune_descendants_of(parent_header.header.hash());
         };
 
         // TODO(ved): derive randomness, solution range, salt if interval is met
         // TODO(ved): extract record roots from the header
         // TODO(ved); extract an equivocations from the header
+        // TODO(ved):
+        //      at the moment, we cannot prune the fork headers due to the probabilistic nature of the chain
+        //      Once we have some form of finality to the chain, we should prune the forks then
```

Comment:
Well, arguably on chain light client that doesn't prune anything isn't great. Fine for this PR, but should be the nearest priority to implement pruning. We are using archiving depth as a finality right now for practical purposes.